staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go

```diff
@@ -238,43 +325,272 @@ func newAuthenticator(opts Options, initVerifier func(ctx context.Context, a *Au
+	// for this claim.
+	verifierPerIssuer map[string]*asyncIDTokenVerifier
+}
+
+func (i claimInfo) String() string {
+	return fmt.Sprintf("name:%q, verifierPerIssuer:%+v", i.name, i.verifierPerIssuer)
+}
+
+// claimResolver expands distributed claims by calling respective claim source
+// endpoints.
+type claimResolver struct {
+	// claimInfos is a map from a claim to verifier map.
+	claimInfos map[string]claimInfo
+
+	// A HTTP transport to use for distributed claims
+	t *http.Transport
+}
+
+// newClaimResolver creates a new resolver for distributed claims.  Each claim
+// has a number of allowed issuers, as specified by issuersPerClaim, and only
+// those issuers are honored.
+// TODO: Support separate client IDs per different issuer/claim.
+func newClaimResolver(ctx context.Context, issuersPerClaim map[string][]string, t *http.Transport, config *oidc.Config) *claimResolver {
+	cp := &claimResolver{claimInfos: map[string]claimInfo{}, t: t}
+	for c, issuers := range issuersPerClaim {
+		verifierPerIssuer := map[string]*asyncIDTokenVerifier{}
+		for _, iss := range issuers {
+			glog.V(10).Infof("adding claim info: %v", iss)
+			verifierPerIssuer[iss] = newAsyncIDTokenVerifier(ctx, config, iss)
+		}
+		ci := claimInfo{c, verifierPerIssuer}
+		cp.claimInfos[c] = ci
+		glog.V(10).Infof("adding claim info: %+v", ci)
+	}
+	return cp
+}
+
+// Verifier returns either the verifier for the specified issuer, or error.
+func (p *claimResolver) Verifier(claim, iss string) (*oidc.IDTokenVerifier, error) {
+	ci, ok := p.claimInfos[claim]
+	if !ok {
+		return nil, fmt.Errorf("no claim info for claim: %q, issuer: %q", claim, iss)
+	}
+	av := ci.verifierPerIssuer[iss]
+	if av == nil {
+		return nil, fmt.Errorf("no verifier for claim: %q, issuer: %q", claim, iss)
+	}
+	v := av.verifier()
+	if v == nil {
+		return nil, fmt.Errorf("verifier not initialized for claim: %q, issuer: %q", claim, iss)
+	}
+	glog.V(10).Infof("Got verifier for: claim: %q, iss: %q: %+v", claim, iss, v)
+	return v, nil
+}
+
+// expand extracts the distributed claims from claim names and claim sources.
+// The extracted claim value is pulled up into the supplied claims.
+//
+// Distributed claims are of the form as seen below, and are defined in the
+// OIDC Connect Core 1.0, section 5.6.2.
+// See: https://openid.net/specs/openid-connect-core-1_0.html#AggregatedDistributedClaims
+//
+// {
+//   ... (other normal claims)...
+//   "_claim_names": {
+//     "groups": "src1"
+//   },
+//   "_claim_sources": {
+//     "src1": {
+//       "endpoint": "https://www.example.com",
+//       "access_token": "f005ba11"
+//     },
+//   },
+// }
+func (r *claimResolver) expand(c claims) error {
+	const (
+		// The claim containing a map of endpoint references per claim.
+		// OIDC Connect Core 1.0, section 5.6.2.
+		claimNamesKey = "_claim_names"
+		// The claim containing endpoint specifications.
+		// OIDC Connect Core 1.0, section 5.6.2.
+		claimSourcesKey = "_claim_sources"
+	)
+
+	if glog.V(11) {
```

Comment:
glog.V(1).Infof(...)

Reply:
I'd rather not --- the claim sets can be very large and even just invoking the logger at the appropriate level carries a performance penalty.

This side-steps the issue by only ever calling Infof if there's a chance that it's actually going to be logged.